sycl/doc/extensions/proposed/sycl_ext_oneapi_usm_properties.asciidoc

```diff
@@ -0,0 +1,169 @@
+= sycl_ext_oneapi_usm_properties
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension introduces an alternative to current usm property list: usm_property_list. This accepts compile-time-constant properties that is also available to runtime, 
+`malloc_device`, `malloc_host`, `malloc_shared` take this new property list and modifies malloc to be templated on its usm_property_list.
+
+The goal of these changes is to enable information about properties to propagate to the device compiler and thereby enable additional optimization of kernel code.
+Further, this will also be used by the runtime for allocation of the memory.
+
+== Notice
+
+Copyright (c) 2021-2022 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a proposed extension specification, intended to gather community
+feedback. Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state. The specification itself may also change in
+incompatible ways before it is finalized. Shipping software products should not
+rely on APIs defined in this specification.
+
+== Version
+
+Revision: 1
+
+== Contributors
+
+Abhishek Tiwari, Intel +
+Aditi Kumaraswamy, Intel +
+Gregory Lueck, Intel +
+Jason Sewall, Intel +
+Jessica Davies, Intel +
+Joe Garvey, Intel +
+Mike Kinsner, Intel +
+Sherry Yuan, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4 and the following extensions:
+
+- link:sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+
+=== Examples
+
+The properties can be used in the usm allocation as follows:
+
+[source,c++]
+----
+
+using namespace sycl::ext::oneapi;
+
+annotated_ptr<int, property_list_t<sycl::ext::intel::buffer_location<1>>> data_v = malloc_device<int, property_list_t<sycl::ext::intel::buffer_location<1>>>(N, q);
```

Comment:
Possible improvements 1: is it possible to make it less verbose, i.e combine the runtime and compile time property, and pass both as template arg of annotated_ptr?

Or is it possible to directly pass [sycl_ext_oneapi_properties](https://github.com/intel/llvm/blob/1df003896532b3aa4454ea5c061eaf9b25ada045/sycl/doc/extensions/proposed/sycl_ext_oneapi_properties.asciidoc) into the function parameter, and then pass it as one of the template arguments for annotated ptr?

Reply:
Hopefully a user could do something like the following:
```
auto data = malloc_device<int>(N, q, property_list{my_runtime_property{4}, sycl::ext::intel::buffer_location_v<1>});
// data is of type annotated_ptr<int, property_list_t<sycl::ext::intel::buffer_location<1>>>
```
The proposal could be:
```
template <typename T>
annotated_ptr<T, PropertyListT> malloc_device(
    size_t Count, const queue &Q, const sycl::ext::oneapi::experimental::properties &PropList = {},
    const detail::code_location CodeLoc = detail::code_location::current()) {
  annotated_ptr<T, PropList> addr(malloc_device<T>(Count, Q.get_device(), Q.get_context(), PropList,
                          CodeLoc));
  return addr;
}
```
I don't think `sycl::ext::oneapi::experimental::properties` support this kind of usage?

or
```
template <typename T>
annotated_ptr<T, PropertyListT> malloc_device(
    size_t Count, const queue &Q, const sycl::ext::oneapi::experimental::properties &PropList = {},
    const detail::code_location CodeLoc = detail::code_location::current()) {
  if (buffer loc is property of PropList)
    annotated_ptr<T, sycl::ext::intel::buffer_location<1>> addr(malloc_device<T>(Count, Q.get_device(), Q.get_context(), PropList, CodeLoc));
  return addr;
}
```
But it has scalability concern, what if there are other properties?

Reply:
Base on: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties.asciidoc#adding-a-property-list-to-a-kernel-launch

another potential template is

```cpp
 template <typename T, typename PropertyList>
annotated_ptr<T, PropertyListT> annotated_malloc_device(
    size_t Count, const queue &Q, PropertyList properties,
    const detail::code_location CodeLoc = detail::code_location::current()) {
    annotated_ptr<T, PropertyList> ptr(addr);
    return addr;
}

sycl::ext::oneapi::experimental::properties properties{sycl::ext::oneapi::buffer_location<2>, runtime_prop(3)};
auto data = malloc_device<int>(N, q, properties);
// data is of type annotated_ptr<int, property_list_t<sycl::ext::intel::buffer_location<1>>>
```

Not sure of correctness yet.